File: codeAccumulate/spark/src/main/scala/com/sunsky/secondSort/SecondarySortApp.scala
package com.sunsky.secondSort

import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created by sunsky on 2017/3/20.
  * first key sort then the value sort
  */
object SecondarySortApp {
  def main(args: Array[String]): Unit = {
    val sc = getContext("Secondary Order")
    secondarySort(sc)
    sc.stop()
  }

  def getContext(name : String) = {
    val conf = new SparkConf().setAppName(name).setMaster("local")
    val sc = new SparkContext(conf)
    sc
  }

  def secondarySort(sc : SparkContext) = {
    val lines = sc.textFile("D://file//sort.dat")
    val pairWithSortkey = lines.map(
      line => (new SecondarySortKey(line.split(" ")(0).toInt,line.split(" ")(1).toInt),line)
    )

    val sorted = pairWithSortkey.sortByKey(false)

    val sortedResult = sorted.map(line => line._2)

    sortedResult.collect().foreach(println)
  }
}

File: codeAccumulate/spark/src/main/java/com/sunsky/javacode/secondSort/SecondarySortApp.java
package com.sunsky.javacode.secondSort;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

/**
 * Created by sunsky on 2017/3/20.
 */
public class SecondarySortApp {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setAppName("second sort");
        conf.setMaster("local");
        JavaSparkContext context = new JavaSparkContext(conf);

        JavaRDD<String> lines = context.textFile("D://file//sort.dat");

        JavaPairRDD<SecondaryOrderKey,String> paris = lines.mapToPair(new PairFunction<String, SecondaryOrderKey, String>() {

            public Tuple2<SecondaryOrderKey, String> call(String line) throws Exception {
                String[] splited = line.split(" ");
                SecondaryOrderKey key = new SecondaryOrderKey(Integer.parseInt(splited[0]), Integer.parseInt(splited[1]));
                return new Tuple2<SecondaryOrderKey, String>(key, line);
            }
        });

        JavaPairRDD<SecondaryOrderKey,String> sorted = paris.sortByKey();

        JavaRDD secondarySorted = sorted.map(new Function<Tuple2<SecondaryOrderKey,String>, String>() {

            public String call(Tuple2<SecondaryOrderKey, String> v1) throws Exception {
                return v1._2;
            }
        });

        secondarySorted.foreach(new VoidFunction<String>(){

            public void call(String t) throws Exception {
                System.out.println(t);
            }

        });
    }
}
